ns for fx: remove restriction on nodes with no args and only kwargs

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#78181 ns for fx: remove restriction on nodes with no args and only kwargs**

Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

vkuzo committed:
ns for fx: remove restriction on nodes with no args and only kwargs

Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: remove restriction on nodes with no args and only kwargs"


Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: remove restriction on nodes with no args and only kwargs"


Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

[ghstack-poisoned]

z-a-f commented on torch/ao/ns/fx/graph_passes.py:
```diff
@@ -27,6 +27,33 @@
+        else:
+            assert len(node.args) > idx
+            return node.args[idx]  # type: ignore[return-value]
+    except RuntimeError:
```
nit: Can we add a comment what use case this exception is catching?

z-a-f reviewed (approved):
sgtm

jerryzh168 commented on torch/ao/ns/fx/graph_passes.py:
```diff
@@ -27,6 +27,33 @@
+            gm, normalize_to_only_use_kwargs=True)
+        if norm_args_and_kwargs is not None:
+            norm_args, norm_kwargs = norm_args_and_kwargs
+            assert len(norm_args) + len(norm_kwargs) >= idx + 1
```
why is this `idx` + 1 and L50 is `idx`? can we have a comment?

jerryzh168 replied:
oh this is `>=` and the other one is `>`, maybe better to use the same?

jerryzh168 commented on torch/ao/ns/fx/graph_passes.py:
```diff
@@ -396,14 +426,40 @@ def _can_insert(node_a_arg, gm_a):
-    for node_a_arg in nodes[0].args[num_non_param_args_node_a:]:
-        if not _can_insert(node_a_arg, gm_a):
-            return False
+    for node_a in nodes:
```
is it possible to have a function to return `nodes[num:]`?

vkuzo replied:
yeah, I looked into it, didn't see a way that was significantly better than the current code

vkuzo committed:
Update on "ns for fx: remove restriction on nodes with no args and only kwargs"


Summary:

Removes the restriction from NS for FX on handling nodes which have
no positional arguments, such as `F.linear(input=x, weight=w, bias=b).

In order to achieve this, we delete all places in the code which
were doing things like

```
node.args[0]
```

And replace them with

```
_get_normalized_nth_input(node, gm, 0)
```

The `_get_normalized_nth_input` function is a best effort way to
get the n'th normalized input.

This is needed because some FX tools output nodes normalized to
be kwargs only, and we need to be able to handle this in NS.

Test plan:

```
python test/test_quantization.py -k test_linear_kwargs_shadow
```

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge this please
